File: src/storage/storage.ts
import AsyncStorage from '@react-native-async-storage/async-storage';
import { ExpenseEntry, SalaryEntry } from '../types';

const monthKey = (month: number) => String(month + 1).padStart(2, '0');

const salaryKey = (businessId: string, year: number, month: number) =>
  `myca:salaries:${businessId}:${year}-${monthKey(month)}`;

const expenseKey = (businessId: string, year: number, month: number) =>
  `myca:expenses:${businessId}:${year}-${monthKey(month)}`;

const safeParse = <T>(value: string | null): T[] => {
  if (!value) {
    return [];
  }

  try {
    return JSON.parse(value) as T[];
  } catch {
    return [];
  }
};

export const getSalaries = async (businessId: string, year: number, month: number): Promise<SalaryEntry[]> => {
  const value = await AsyncStorage.getItem(salaryKey(businessId, year, month));
  return safeParse<SalaryEntry>(value);
};

export const saveSalaries = async (
  businessId: string,
  year: number,
  month: number,
  entries: SalaryEntry[],
): Promise<void> => {
  await AsyncStorage.setItem(salaryKey(businessId, year, month), JSON.stringify(entries));
};

export const getExpenses = async (businessId: string, year: number, month: number): Promise<ExpenseEntry[]> => {
  const value = await AsyncStorage.getItem(expenseKey(businessId, year, month));
  return safeParse<ExpenseEntry>(value);
};

export const saveExpenses = async (
  businessId: string,
  year: number,
  month: number,
  entries: ExpenseEntry[],
): Promise<void> => {
  await AsyncStorage.setItem(expenseKey(businessId, year, month), JSON.stringify(entries));
};

File: MyCA/MyCA/Storage/Storage.swift
import Foundation

struct Storage {
    private static let defaults = UserDefaults.standard

    static func salaries(businessId: String, year: Int, month: Int) -> [SalaryEntry] {
        load(forKey: key(prefix: "myca.salaries", businessId: businessId, year: year, month: month))
    }

    static func saveSalaries(_ entries: [SalaryEntry], businessId: String, year: Int, month: Int) {
        save(entries, forKey: key(prefix: "myca.salaries", businessId: businessId, year: year, month: month))
    }

    static func expenses(businessId: String, year: Int, month: Int) -> [ExpenseEntry] {
        load(forKey: key(prefix: "myca.expenses", businessId: businessId, year: year, month: month))
    }

    static func saveExpenses(_ entries: [ExpenseEntry], businessId: String, year: Int, month: Int) {
        save(entries, forKey: key(prefix: "myca.expenses", businessId: businessId, year: year, month: month))
    }

    private static func key(prefix: String, businessId: String, year: Int, month: Int) -> String {
        String(format: "%@.%@.%04d-%02d", prefix, businessId, year, month)
    }

    private static func load<T: Decodable>(forKey key: String) -> [T] {
        guard let data = defaults.data(forKey: key) else { return [] }
        return (try? JSONDecoder().decode([T].self, from: data)) ?? []
    }

    private static func save<T: Encodable>(_ values: [T], forKey key: String) {
        if let data = try? JSONEncoder().encode(values) {
            defaults.set(data, forKey: key)
        }
    }
}
